### PR TITLE
[WIP] link: sync deps with parent package

### LINKED ISSUE #46: Synchronize dependencies of a `gx-go link` package with another one to avoid discrepancies
As discussed in https://github.com/ipfs/go-unixfs/issues/4 (the main scenario being: I'm working on a git-cloned `go-unixfs` repo and I want to build it against `go-unixfs` so I `gx-go link` it but then there are dependency discrepancies), when using `gx-go link` the dependencies mismatch don't allow to build the `go-ipfs` repo against the linked dependency.

As a temporary solution I would like to add an (experimental) feature in the `gx-go rewrite` command to use the dependencies of a parent repository (e.g., `go-ipfs`) instead of its own when there's a dependency mismatch. This is somewhat of an already implicit (unpremeditated) functionality, as rewrite map collisions are already handled by overwriting themselves (https://github.com/whyrusleeping/gx-go/issues/44).

The tricky part is how to handle the reverse (`--undo`) process. Since we may have built the rewrite map using dependencies of another package, if that package is modified (some of its dependencies changed/updated), the old rewritten dependencies may not match any of the dependencies in the new rewrite map and they won't get reverted (I'm not sure if the `--fix` functionality can be of use here).

So, a few options, keeping in mind that this is (hopefully) a  temporary solution while bigger design decisions are being made in https://github.com/whyrusleeping/gx/issues/179:

1. Don't bother with the un-rewrite process, since this feature is intended to be used with the `gx-go link` command over git-controlled repos, so the rewrite can be reverted with git.

2. Do the `rewrite --undo`, but leave the burden on the user to keep track of which version of the parent package it used to sync the dependencies.

3. (My personal favorite.) Save the rewrite map (something that I think would be desirable in many scenarios besides this one) and revert the imports based on that and do not re-read the linked package dependencies or any other parent package used to sync them. This could be save as a JSON file under the `.gx/` directory.

### PR DESCRIPTION
**Pull request by schomatis**
This PR extracts the `--sync` feature from https://github.com/whyrusleeping/gx-go/pull/48 without the `vendor/` support (that has some implementation problems, see https://github.com/whyrusleeping/gx-go/issues/46#issuecomment-416229688).

As the imported PR, this command now requires `gx-go link` to be run inside the parent package of the dependency even though the link is still done to the global workspace. This breaks API but it makes easier to reason about the code and about the link being done, and in most use cases the developer was already calling the command from a package.

Closes #46.

### CODE REVIEW

**Comment by schomatis** [comment]
(Another possible enhancement for this PR or a new one: detect already linked packages and un-link+link them again. At the moment it would seem that the `post-install` hook + rewrite functionality doesn't re-process an already rewritten file -it's assuming that the import paths wont change?, but if we've updated something?- so if I link without `--sync` and it fails due to a mismatch version, when I link again, now with `--sync`, the rewrite part of the process won't happen but it won't fail either. The result is that both with and without the sync functionality will seem to fail -the user will need to actually un-link and re-link manually, but this won't be obvious to him/her-.)

**Comment by Stebalien** [comment]
> detect already linked packages and un-link+link them again.

We *should* probably do that but we can do it in a new patch.

File: link.go [hunk 1]
```diff
@@ -19,30 +19,34 @@ var LinkCommand = cli.Command{
 	Usage: "Symlink packages to their dvcsimport repos, for local development.",
 	Description: `gx-go link eases local development by symlinking actual workspace repositories on demand.
 
-Example workflow:
+Links the specified dependency package (by name or hash) used by the
+parent package (in the CWD or specified by --parent). The link is done
```
[the remark this answers is not in the document]
Comment (schomatis) [reply link.go]: I'd prefer to maintain the parent-child relationship concept, because everything is a package here  *but*, depending on where you're looking from (in what part of the code you're in), that package is either a dependency (and that distinction is important because a dependency in a `package.json` doesn't have the same information as the actual package) or the package that's importing that dependency. The term "current" doesn't seem to carry that relationship meaning, is there another term you'd prefer?

File: link.go [hunk 2]
```diff
@@ -53,21 +57,33 @@ unlinked QmVGtdTZdTFaLsaj2RwdVG8jcjNNcp1DE914DKZ2kHmXHw /home/user/go/src/github
 			Name:  "a,all",
 			Usage: "Remove all existing symlinks and reinstate the gx packages. Use with -r.",
 		},
+		cli.StringFlag{
+			Name:  "p,parent",
+			Usage: "Specify the path of the parent package of the linked dependency (default: CWD).",
+		},
+		cli.BoolFlag{
+			Name:  "s,sync",
```
[the remark this answers is not in the document]
Comment (schomatis) [reply link.go]: Yes, I don't really like "sync" but couldn't come up with something better. Let's go with `--deps=` and use the term that will replace "parent" here, e.g., `--deps=<package-that-is-including-this-package>`.

File: link.go [hunk 3]
```diff
@@ -53,21 +57,33 @@ unlinked QmVGtdTZdTFaLsaj2RwdVG8jcjNNcp1DE914DKZ2kHmXHw /home/user/go/src/github
...
+			Usage: "Synchronize the dependencies of the linked package with the ones of its parent.",
```
[the remark this answers is not in the document]
Comment (schomatis) [reply link.go]: Would you prefer to use "dependent"?

File: link.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply link.go]: When I'm using GX, I'm usually working *on* a gx package. The "current" package is the package I'm currently working on.

For example:

> `gx-go link` replaces the target gx package (either by name or hash) with a symlink to the appropriate dvcs repo in your GOPATH. To make this "symlinked" repo usable as a gx package, `gx-go link` rewrites the target package's dvcs imports using the target package's `package.json`. Unfortunately, this can cause build errors in packages *depending* on this package if these dependent packages specify alternative, conflicting dependency versions. We can work around this by asking `gx` to rewrite the target package using dependencies from the current package (the package you're trying to build) first, falling back on the target package's `package.json` file. We can do this by passing the `--deps=mine` flag.

That *is* a bit wordy and you're right, "current" may not be quite right.

File: link.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply link.go]: I would, but then you have to answer the question "which dependent". The answer is "the current one".

**Comment by Stebalien** [comment]
Thinking about this, we could actually use the dependencies specified in the *original* gxed package (the one we're replacing with a symlink). That would be the most conservative approach (effectively "downgrading" dependencies.

The only difference is that, if the target repo has added a new dependency *also* present in the current/parent package, we'll use the version in the *target* repo, not the current/parent. However, using the package.json of the current/parent isn't foolproof either (e.g., the target could have added a dependency that we *don't* have, one that depends on packages we *do* depend on but depends on different versions).


The other approach would be to do a fake gx-workspace update. However, that's *way* more work.

File: link.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (schomatis) [reply link.go]: Let's leave it at "the current dependent package". Nice text (I like it *wordy* :), I'll add it to the usage doc.

**Comment by schomatis** [comment]
> The only difference is that, if the target repo has added a new dependency also present in the current/parent package, we'll use the version in the target repo, not the current/parent. However, using the package.json of the current/parent isn't foolproof either (e.g., the target could have added a dependency that we don't have, one that depends on packages we do depend on but depends on different versions).

Let's have both options: `--deps=dependent` and `--deps=dependency` (we can use other terms but let's keep them absolute, not relative to the current/mine/theirs perspective) and recommend the second one (and have it as the default) as the conservative approach (noting that it not foolproof).

**Comment by Stebalien** [comment]
> Let's have both options: --deps=dependent and --deps=dependency (we can use other terms but let's keep them absolute, not relative to the current/mine/theirs perspective) and recommend the second one (and have it as the default) as the conservative approach (noting that it not foolproof).

In this case, the other strategy would really be "original" or "replaced".

I'm really confused why you feel so strongly about keeping the terminology absolute. Personally, I find it *extremely* confusing (depending on the context). When I see an absolute term, I think "oh, they must *not* mean my current package, otherwise they would have just said that". Specifically, when I see "dependent", my first thought is "which dependent?".

**Comment by schomatis** [comment]
> I'm really confused why you feel so strongly about keeping the terminology absolute. Personally, I find it extremely confusing (depending on the context).

Ok, let me rethink that and get back to you.

I think that my main objection is about the code and not the UX. From the point of the user who just runs `cd package; gx-go link dependency` I think you're right, "current" is the package we just moved to, there's no doubt there. Using "current" in the code starts losing meaning as we're constantly changing perspectives between dependency-dependent packages, and the original ("current") package from where we (the user) started has little to do with the ongoing operation of, for example, fetching the dependency of the the dependency. That's where I want to have a clear model of which package is on top from where we are extracting its dependencies to later convert (fetch) them into other (also) packages. But yes, that part of the code is a different realm from the command the user runs.

**Comment by Stebalien** [comment]
Ah, yes, *that's* totally reasonable. `parent` actually seems like a good word in that case (or `root`).

**Comment by schomatis** [comment]
@Stebalien Could you take another look please? Changed the terminology to target/current (using your command description) and renamed the option `--override-deps`. I've finally had to settle for just using the versions from the current dependent package (and didn't add the variation of taking them from the target package because this command overwrites it -since everything is done in the same global directory- and wanted to avoid changing too much functionality in this first iteration).

**Comment by schomatis** [comment]
@Stebalien

**Comment by schomatis** [comment]
@Stebalien
